torch_glow/src/PyTorchModelLoader.cpp

```diff
@@ -25,47 +28,354 @@ glow::Type ptTypeToGlowType(const at::CompleteTensorType &ptType) {
+      "conv_input_transposed", conv->getResult(), NHWC2NCHW);
+
+  addGlowNodeValue(outputs[0], output->getResult());
+}
+
+void PyTorchModelLoader::loadConstant(const torch::jit::Node *ptNode) {
+  auto inputs = ptNode->inputs();
+  auto outputs = ptNode->outputs();
+  assert(inputs.size() == 0);
+  assert(outputs.size() == 1);
+
+  auto optionalIValue = torch::jit::toIValue(outputs[0]);
+  assert(optionalIValue.has_value() && "Constants should have IValue outputs");
+  const torch::jit::IValue iVal = *optionalIValue;
+
+  glow::Module *mod = f_->getParent();
+
+  glow::Tensor t;
+
+  if (iVal.isInt()) {
+    int64_t val = iVal.toInt();
+    t.reset(glow::ElemKind::Int32ITy, {1});
+    t.getHandle<int32_t>().raw(0) = to32Bit(val);
+  } else if (iVal.isIntList()) {
+    const auto vals = iVal.toIntListRef();
+    t.reset(glow::ElemKind::Int32ITy, {vals.size()});
+    auto handle = t.getHandle<int32_t>();
+    for (size_t i = 0; i < vals.size(); ++i) {
+      handle.raw(i) = to32Bit(vals[i]);
+    }
+  } else if (iVal.isDouble()) {
+    double val = iVal.toDouble();
+    t.reset(glow::ElemKind::FloatTy, {1});
+    t.getHandle<float>().raw(0) = to32Bit(val);
+  } else if (iVal.isDoubleList()) {
+    const auto vals = iVal.toDoubleListRef();
+    t.reset(glow::ElemKind::FloatTy, {vals.size()});
+    auto handle = t.getHandle<float>();
+    for (size_t i = 0; i < vals.size(); ++i) {
+      handle.raw(i) = to32Bit(vals[i]);
+    }
+  } else if (iVal.isBool()) {
+    bool val = iVal.toBool();
+    t.reset(glow::ElemKind::BoolTy, {1});
+    t.getHandle<bool>().raw(0) = val;
+  } else if (iVal.isBoolList()) {
+    const auto &valsList = iVal.toBoolList();
+    const auto &vals = c10::impl::toVector(valsList);
+    t.reset(glow::ElemKind::BoolTy, {vals.size()});
+    auto handle = t.getHandle<bool>();
+    for (size_t i = 0; i < vals.size(); ++i) {
+      handle.raw(i) = vals[i];
+    }
+  } else if (iVal.isTensor()) {
+    assert(false && "not yet implemented tensor constants");
+  } else if (iVal.isNone()) {
+    // Nothing to do for None
+    return;
+  } else {
+    assert(false && "Unsupported constant type");
   }
+
+  glow::Constant *glowConstant = mod->createConstant("constant", std::move(t));
+  addGlowNodeValue(outputs[0], glowConstant->getOutput());
+}
+
+void PyTorchModelLoader::loadNode(const torch::jit::Node *ptNode) {
+  switch (ptNode->kind()) {
+  case at::prim::Constant:
+    return loadConstant(ptNode);
+  case at::aten::mul:
+    return loadMul(ptNode);
+  case at::aten::div:
+    return loadDiv(ptNode);
+  case at::aten::add:
+    return loadAdd(ptNode);
+  case at::aten::sub:
+    return loadSub(ptNode);
+  case at::aten::_convolution:
+    return loadConvolution(ptNode);
+  case at::aten::relu:
+    return loadRelu(ptNode);
   default:
     assert(false && "unrecognized node");
   }
 }
 
+// static
+bool PyTorchModelLoader::isNodeSupported(const torch::jit::Node *node) {
+  switch (node->kind()) {
```

Comment:
constant?

Reply:
what do you mean?

Reply:
Oh Constant node? That's a good point, I meant to add a comment for this. So we don't want to actually add it here or else I think pt will hand over all constants to glow because glow says it can support them. If we leave it out then pt hands glow the constants that it needs because it has to express them somehow in the subgraph that gets handed to glow

Reply:
Agreed, a little different from XXXFill ops in c2.